### DISCUSSION TITLE
Settling time for spring/damper?

### DISCUSSION THREAD

**Discussion by salk31** [dyn4j/dyn4j#223] [Q&A] [answered]
Hi,

I'm still, slowly, trying to simulate different spring counter balance mechanisms. I've got as far as a mass pivoted at one end using a RevoluteJoint/motor and a spring pulling to balance out the mass (for a given K and rest distance I calculate what the current distance should be). The RevoluteJoint and DistanceJoint just attaching to ground.

It seems OK except getReactionTorque on the motor called during postSolve seems to take a while to settle to what I think is the correct value. e.g. if I set the motor speed to zero the reactionTorque starts at 3ish and takes a hundred or so iterations to get near to zero. 

Any obvious reasons for this? Starting the spring/damper not at restDistance? Do I need to fiddle with the duration of the time step?

NB I also use another body and WeldJoint to give the spring leverage (not the simplest way of doing it but I want multiple levers later).

Thanks

Sam

**Comment by wnbittle** [comment]
I'm not sure I fully understand the scenario, but I would expect that if a spring and motor are counter balanced that their reaction force/torque would oscillate then settle after a number of iterations.  Visually are the objects in equilibrium?  It's possible that the solver counters one, then counters the other, and so on until equilibrium is reached, all while visually appearing in equilibrium.

**Reply by salk31** [reply]
I'm not sure that fits. It looks more like exponential decay rather than oscillation. No visible sign of movement. If I clean up the source enough to be readable would you have time to look? 

Either way I appreciated your help and all the work on dyn4j.

**Reply by wnbittle** [reply]
I'm not surprised that the reaction torque/force would converge to a value over time since that's how the solver is designed.  I am a little surprised there's not some oscillation though.  Regarding how long it takes to converge... if I understand the scenario that's also not too surprising since the solver basically gives up as soon as it can to improve performance.  The only thing I can think of is to fiddle with the number of solver iterations (velocity) - increase it to something higher like 20 or 50 and see if convergence is any better.

If it's not too much trouble I'd like to see a sample of the behavior.  Do you have any output that shows the reaction torque/force over time (one value per frame or something)?  When you say exponential decay, can you confirm that there's convergence to a value and that it's strictly decreasing?  The time it takes to converge - can you clarify 300 frames or 300 iterations (considering that each timestep performs n iterations)?

**Reply by salk31** [reply]
Sorry, no. Not strictly decreasing. If I get the time tomorrow I'll try and do a chart. I've attached raw data [v1.csv](https://github.com/dyn4j/dyn4j/files/8772418/v1.csv).

Here is my minimal-ish example. It still has some nonsense for drawing levers and calculating the K and rest length of my real springs.


https://github.com/salk31/dyn4j-samples/blob/master/src/main/java/org/dyn4j/samples/SpringBalance.java

(edit 1 - sorry about copy n pasting, I didn't realise it was so much and I thought github UI would do something clever)

**Comment by wnbittle** [comment]
Sorry for such a delay here, I am looking into this, but life keeps coming up.  I'll take another look tonight.

**Reply by salk31** [reply]
I'm very grateful that you are looking at all. I think I know what you mean, with work, Ukrainian guests, chronic illness I've not been able to poke at my hobby for a while ;) I have been day dreaming about using high school level physics to calculate the forces in parallel though.

**Comment by wnbittle** [comment, marked as answer]
I've discovered the issue - the problem is that the `createRod` method returns a fixture with a very low density (0.001).  When I change the density to 1 it converges quickly. Is there a reason the density is so low for that shape?

The reason this happens is due to the solver.  It's very good at what it does, but large mass ratios are it's arch enemy.

**Reply by salk31** [reply]
Many thanks! Is it a numerical accuracy issue? I only made it low density so it wouldn't impact the balancing. I think I noticed that very low frequency and damping of the springs made it unhappy, similar issue?

Anyway, many thanks for taking the time to look at that. I'm not sure I'd ever have got there.

**Comment by wnbittle** [comment]
The issue is that large differences in mass cause the solver to converge very very slowly.  Take a look at [this - slide 30](https://box2d.org/files/ErinCatto_UnderstandingConstraints_GDC2014.pdf).  The results there indicated that to converge the `velocity solver iterations = 3 * (mass ratio) + 2`.  Note that this is in the context of collision resolution, but would apply to joints as well because they run through the same solver.

Modifying the settings like `this.world.getSettings().setVelocityConstraintSolverIterations(1000);` causes it to converge quickly with your original rod mass of 0.001.  For comparison, the default value for this is `10`, so this can really impact performance, but it's probably not noticable in a small simulation like your example.

As to why it converges slowly: The mass of the rod compared to the other object is low and causes the majority of the impulse to resolve the pair-wise constraint to be applied to the rod, not to the load.  As a result, solving the _system_ takes forever because solving the system really means applying more impulse to the load (the larger mass body).
